orchagent/port_rates.lua

```diff
@@ -29,11 +32,117 @@ logit(alpha)
+
+-- lookup interface name from port oid
+
+local function find_interface_name_from_oid(port)
+
+    for i = 1, port_interface_oid_key_count do
+        local index = i * 2 - 1
+        if port_interface_oid_map[index + 1] == port then
+            return port_interface_oid_map[index]
+        end
+    end
+
+    return 0
+end
+
+-- calculate lanes and serdes speed from interface lane count & speed
+-- return lane speed and serdes speed
+
+local function calculate_lane_and_serdes_speed(count, speed)
+
+   local serdes = 0
+   local lane_speed = 0
+
+    if count == 0 or speed == 0 then
+        logit("Invalid number of lanes or speed")
+        return 0, 0
+    end
+
+    -- check serdes_cnt if it is a multiple of speed
+    local serdes_cnt = math.fmod(speed, count)
+
+    if serdes_cnt ~= 0 then
+        logit("Invalid speed and number of lanes combination")
+        return 0, 0
+    end
+
+    lane_speed = math.floor(speed / count)
+
+    -- return value in bits
+    if lane_speed == 1000 then
+        serdes = 1.25e+9
+    elseif lane_speed == 10000 then
+        serdes = 10.3125e+9
+    elseif lane_speed == 25000 then
+        serdes = 25.78125e+9
+    elseif lane_speed == 50000 then
+        serdes = 53.125e+9
+    elseif lane_speed == 100000 then
+        serdes = 106.25e+9
+    else
+       logit("Invalid serdes speed")
+    end
+
+    return lane_speed, serdes
+end
+
+-- look up interface lanes count, lanes speed & serdes speed
+-- return lane count, lane speed, serdes speed
+
+local function find_lanes_and_serdes(interface_name)
+    -- get the port config from config db
+    local _
+    local serdes, lane_speed, count = 0, 0, 0
+
+    -- Get the port configure
+    redis.call('SELECT', appl_db)
+    local lanes = redis.call('HGET', appl_db_port ..':'..interface_name, 'lanes')
+
+    if lanes then
+        local speed = redis.call('HGET', appl_db_port ..':'..interface_name, 'speed')
+
+        -- we were spliting it on ','
+        _, count = string.gsub(lanes, ",", ",")
+        count = count + 1
+
+        lane_speed, serdes = calculate_lane_and_serdes_speed(count, speed)
+
+    end
+    -- switch back to counter db
+    redis.call('SELECT', counters_db)
+
+    return count, lane_speed, serdes
+end
+
 local function compute_rate(port)
+
     local state_table = rates_table_name .. ':' .. port .. ':' .. 'PORT'
     local initialized = redis.call('HGET', state_table, 'INIT_DONE')
     logit(initialized)
 
+    -- FEC BER
+    local fec_corr_bits, fec_uncorr_frames
+    local fec_corr_bits_ber_new, fec_uncorr_bits_ber_new = -1, -1
+    -- HLD review suggest to use the statistical average when calculate the post fec ber
+    local rs_average_frame_ber = 1e-8
+    local lanes_speed, serdes_speed, lanes_count = 0, 0, 0
+
+    -- lookup interface name from oid
+    local interface_name = find_interface_name_from_oid(port)
+    if interface_name then
+        lanes_count, lanes_speed, serdes_speed = find_lanes_and_serdes(interface_name)
```

Comment:
@vincentpcng i was wondering if serdes speed could be populated by OA in APPL_DB port table so that we don't have to compute every time.....may be an optimization for the future

Reply:
In the future if we can add the serdes speed to the appl_db then we can save this calculation.
Agree that we can do it as a future enhancement